src/data_utils.py

```diff
@@ -207,34 +257,92 @@ def _fetch_librispeeh_txt(self) -> dict:
                 all_samples.update(samples)
         return all_samples
 
-    def _fetch_and_push_files(self, data_dir, file_paths: list, file_pattern: str):
-        """All files will be recursively collected from the `data_dir`."""
-        listdir = os.listdir(data_dir)
-        for f in listdir:
-            f = os.path.join(data_dir, f)
-            if f.endswith(file_pattern):
-                f = os.path.abspath(f)
-                file_paths.append(f)
-                continue
 
-            if os.path.isdir(f):
-                self._fetch_and_push_files(f, file_paths, file_pattern)
+class TimitDataLoader(CommonDataLoader):
+    def __init__(self, args: TimitDataLoaderArgs):
+        super().__init__(
+            args.batch_size,
+            args.buffer_size,
+            args.audio_pad_id,
+            args.labels_pad_id,
+            args.audio_maxlen,
+            args.labels_maxlen,
+        )
+
+        self.data_dir = args.data_dir
+
+        self.wav_ext = ".WAV"
+        self.txt_ext = ".TXT"
+
+    def __call__(self, seed=None, drop_remainder=True):
+        wav_files, txt_files = [], []
+        self._fetch_and_push_files(self.data_dir, wav_files, self.wav_ext)
+        self._fetch_and_push_files(self.data_dir, txt_files, self.txt_ext)
+
+        wav_files = set([f[: -len(self.wav_ext)] for f in wav_files])
+        txt_files = set([f[: -len(self.txt_ext)] for f in txt_files])
+
+        # consider only those files which has both text & speech
+        files = list(wav_files & txt_files)
+        print(f"found {len(files)} samples in {self.data_dir}")
+
+        wav_files = [f + self.wav_ext for f in files]
+        txt_files = [f + self.txt_ext for f in files]
+
+        labels = [self._prepare_labels(self.read_timit_txt(f)) for f in txt_files]
+
+        dataset = tf.data.Dataset.from_tensor_slices((wav_files, labels))
+        dataset = dataset.map(
+            lambda sound_path, label: (self.read_sound(sound_path), label)
+        )
+        return self.batchify(dataset, seed=seed, drop_remainder=drop_remainder)
+
+    def _prepare_labels(self, text: str):
+        def _pad(sample: list):
+            while len(sample) < max_length:
+                sample.append(pad_id)
+            return sample
+
+        max_length = self.labels_maxlen
+        pad_id = self.labels_pad_id
+        # TODO : add some processing for cleaning text a bit
+        return _pad(self.tokenizer(text))
+
+    def read_timit_txt(self, file_path: str):
+        with open(file_path, "r") as f:
+            text = f.read().split()[2:]
+        return " ".join(text)
+
+    def read_sound(self, file_path):
+        audio = tf.io.read_file(file_path)
+        audio, _ = tf.audio.decode_wav(audio)
+        return self.processor(tf.squeeze(audio))
 
 
 if __name__ == "__main__":
```

Comment:
Consider turning this into a real test file. You could copy a small sample of files from the dataset to load in tests and verify that the expected data is loaded/